### PR TITLE
Contributions in angle plane-plane and intersection polyline - plane

### PR DESCRIPTION
**Pull request by nikeftekhar**
angle plane plane function
intersection polyline - plane function

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [x] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/compas/geometry/_intersections/intersections.py [hunk 1]
```diff
@@ -383,6 +384,43 @@ def intersection_segment_plane(segment, plane, tol=1e-6):
...
+        polyline to test intersection
+    plane : compas.Geometry.Plane
+        plane to compute intersection
+    tol : float, optional
+        A tolerance for membership verification.
+        Default is ``1e-6``.
+
+    Returns
+    -------
+    intersection_pts : list of compas.geometry.Point
```
Comment (gonzalocasas) [inline src/compas/geometry/_intersections/intersections.py]:
```suggestion
    list of :class:`compas.geometry.Point`
```

**Review by gonzalocasas — COMMENTED** [review]
Generally looks good to me, however, again, if possible, add a couple of unit tests.

File: src/compas/geometry/_core/angles.py [hunk 2]
```diff
@@ -353,6 +354,34 @@ def angles_points_xy(a, b, c, deg=False):
...
 
+def angle_planes(a, b, deg=False):
+    """Compute the smallest angle between the two normal vectors of two planes.
+
+    Parameters
+    ----------
+    a : tuple of a point and a vector. Point as a sequence of float(XYZ) coordinates. Vector as a sequence of float(XY(Z)) coordinates.
+        first plane
+    b : tuple of a point and a vector. Point as a sequence of float(XYZ) coordinates. Vector as a sequence of float(XY(Z)) coordinates.
+        second plane
```
Comment (tomvanmele) [inline src/compas/geometry/_core/angles.py]: this should be
```
a : point and vector or :class:`compas.geometry.Plane`
    The first plane.
b : point and vector or :class:`compas.geometry.Plane`
    The second plane.

File: src/compas/geometry/_core/angles.py [hunk 3]
```diff
@@ -353,6 +354,34 @@ def angles_points_xy(a, b, c, deg=False):
...
+    deg : boolean
+        returns angles in degrees if True
```
Comment (tomvanmele) [inline src/compas/geometry/_core/angles.py]: please add capital letter at the start and period at the end.
angles => angle
True => ``True``

File: src/compas/geometry/_core/angles.py [hunk 4]
```diff
@@ -353,6 +354,34 @@ def angles_points_xy(a, b, c, deg=False):
...
+    Returns
+    -------
+    float
+        The smallest angle in radians (in degrees if deg == True).
+        The angle is always positive.
+
+    Examples
+    --------
+    plane_a = ([0.0, 0.0, 0.0], [0.0, 0.0, 1.0])
+    plane_b = ([0.0, 0.0, 0.0], [1.0, 0.0, 0.0])
```
Comment (tomvanmele) [inline src/compas/geometry/_core/angles.py]: outer parenthesis are not necessary

File: src/compas/geometry/_intersections/intersections.py [hunk 5]
```diff
@@ -383,6 +384,40 @@ def intersection_segment_plane(segment, plane, tol=1e-6):
     return None
 
 
+def intersection_polyline_plane(polyline, plane, expected_number_of_intersections=None, tol=1e-6):
```
Comment (tomvanmele) [inline src/compas/geometry/_intersections/intersections.py:387]: expected number of intersections is not documented

File: src/compas/geometry/_intersections/intersections.py [hunk 6]
```diff
@@ -383,6 +384,40 @@ def intersection_segment_plane(segment, plane, tol=1e-6):
...
+    """Calculate the intersection point of a plane with a polyline. Reduce expected_number_of_intersections to speed up.
+
+    Parameters
+    ----------
+    polyline : compas.geometry.Polyline
```
Comment (tomvanmele) [inline src/compas/geometry/_intersections/intersections.py]:
```
:class:`compas.geometry.Polyline` or sequence of points
```

File: src/compas/geometry/_intersections/intersections.py [hunk 7]
```diff
@@ -383,6 +384,40 @@ def intersection_segment_plane(segment, plane, tol=1e-6):
...
+        polyline to test intersection
+    plane : compas.Geometry.Plane
```
Comment (tomvanmele) [inline src/compas/geometry/_intersections/intersections.py]:
```
:class:`compas.geometry.Plane` or point and vector

File: src/compas/geometry/_intersections/intersections.py [hunk 8]
```diff
@@ -383,6 +384,40 @@ def intersection_segment_plane(segment, plane, tol=1e-6):
...
+        plane to compute intersection
+    tol : float, optional
+        A tolerance for membership verification.
+        Default is ``1e-6``.
+
+    Returns
+    -------
+    list of :class:`compas.geometry.Point`
+        if there are intersection points, return point(s) in a list
```
Comment (tomvanmele) [inline src/compas/geometry/_intersections/intersections.py]: capital and period

File: src/compas/geometry/_intersections/intersections.py [hunk 9]
```diff
@@ -383,6 +384,40 @@ def intersection_segment_plane(segment, plane, tol=1e-6):
...
+        polyline to test intersection
+    plane : compas.Geometry.Plane
+        plane to compute intersection
+    tol : float, optional
+        A tolerance for membership verification.
+        Default is ``1e-6``.
+
+    Returns
+    -------
+    list of :class:`compas.geometry.Point`
```
Comment (tomvanmele) [inline src/compas/geometry/_intersections/intersections.py]: this is not correct.
functions return native Python objects (not COMPAS objects)
```
list of points
```

File: src/compas/geometry/_intersections/intersections.py [hunk 10]
```diff
@@ -383,6 +384,40 @@ def intersection_segment_plane(segment, plane, tol=1e-6):
...
+    for segment in pairwise(polyline):
+        pt = intersection_segment_plane(segment, plane, tol)
+        if pt and max_iter < expected_number_of_intersections:
+            intersection_points.append(pt)
+            max_iter += 1
+        else:
+            break
+    if len(intersection_points) > 0:
+        return intersection_points
+    return None
```
Comment (tomvanmele) [inline src/compas/geometry/_intersections/intersections.py]: this should always return a list, even when it is empty. the truth value of an empty list is the same as of `None`.
```python
return intersections
```

File: src/compas/geometry/_intersections/intersections.py [hunk 11]
```diff
@@ -383,6 +384,40 @@ def intersection_segment_plane(segment, plane, tol=1e-6):
...
+        expected_number_of_intersections = len(polyline)
+    intersection_points = []
+    max_iter = 0
+    for segment in pairwise(polyline):
+        pt = intersection_segment_plane(segment, plane, tol)
+        if pt and max_iter < expected_number_of_intersections:
+            intersection_points.append(pt)
+            max_iter += 1
+        else:
+            break
```
Comment (tomvanmele) [inline src/compas/geometry/_intersections/intersections.py]: why is this not just
```python
for segment in pairwise(polyline):
    if len(intersections) == expected_number_of_intersections:
        break
    point = intersection_segment_plane(segment, plane, tol)
    if point:
        intersections.append(point)
```

Comment (nikeftekhar) [reply src/compas/geometry/_intersections/intersections.py]: yes, you are completely wright. Way better as you indicated. Thank you!

File: src/compas/geometry/_core/angles.py [hunk 12]
```diff
@@ -374,8 +374,8 @@ def angle_planes(a, b, deg=False):
 
     Examples
     --------
-    plane_a = ([0.0, 0.0, 0.0], [0.0, 0.0, 1.0])
-    plane_b = ([0.0, 0.0, 0.0], [1.0, 0.0, 0.0])
+    plane_a = [0.0, 0.0, 0.0], [0.0, 0.0, 1.0]
+    plane_b = [0.0, 0.0, 0.0], [1.0, 0.0, 0.0]
```
Comment (tomvanmele) [inline src/compas/geometry/_core/angles.py:378]: didn't notice this before.
please add `>>>`

**Review by tomvanmele — COMMENTED** [review]
small final stuff...

**Pull request merged**